### PR TITLE
Add StringConstraints for use as Annotated metadata

### PR DESCRIPTION
**Pull request by adriangb**
Related to #6506

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by adriangb** [comment]
please review

**Comment by adriangb** [comment]
ready now :)

File: pydantic/types.py [hunk 1]
```diff
@@ -264,6 +267,52 @@ def conbytes(
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ STRING TYPES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 
+@_dataclasses.dataclass(frozen=True)
+class StringConstraints(annotated_types.GroupedMetadata):
+    """Apply constraints to `str` types.
+
+    Args:
```
Comment (hramezani) [inline pydantic/types.py:274]:
```suggestion
    Attributes:
```

File: pydantic/types.py [hunk 2]
```diff
@@ -264,6 +267,52 @@ def conbytes(
...
+        strip_whitespace: Whether to strip whitespace from the string.
+        to_upper: Whether to convert the string to uppercase.
+        to_lower: Whether to convert the string to lowercase.
+        strict: Whether to validate the string in strict mode.
+        min_length: The minimum length of the string.
+        max_length: The maximum length of the string.
+        pattern: A regex pattern that the string must match.
+
+    Returns:
+        A `StringConstraints` for use in
```
Comment (hramezani) [inline pydantic/types.py:284]:
```suggestion
```

**Review by hramezani — APPROVED** [review]

**Comment by samuelcolvin** [comment]
Should we add documentation now?

**Comment by adriangb** [comment]
> Should we add documentation now?

Yes good call

**Comment by adriangb** [comment]
please review

**Comment by andrewsw** [comment]
> > Should we add documentation now?
> 
> Yes good call

Are y'all aware you've documented the feature but not released it? I just spent an hour trying to figure out why I can't import `StringConstraints` on pydantic version 2.0.3.

**Pull request merged**